### ISSUE TITLE
Takes 30+ seconds to be told 404 for a non existing page

### ISSUE THREAD

**Issue by maxandersen** [openshiftio/openshift.io#3171]
https://github.com/openshiftio/openshift.io/issues/2731 is now fixed which is great; but before I fairly fast got a 404 page I'm now having to wait 30+ seconds.

https://prod-preview.openshift.io/does/not/exist was what I tried as seems #2731 is not fully deployed yet.

**Comment by jiekang** [comment]
One part of this is that visiting a page via the browser url constitutes visiting a 'new' page, resulting in the initial loading of the entire angular application. This is a good number of seconds. Apart from cutting down the load time, I'm not sure we can do much about it. 

Another part is that certain parts of the url are variable, representing the user name or the space name, etc., so a request needs to be made for each of these to see if these are reachable. This takes some time. I wonder if it short-circuits the checks though.

@andrewazores do you have any further thoughts?

**Comment by andrewazores** [comment]
I can't see how the commit that went in (https://github.com/fabric8-ui/fabric8-ui/pull/2686) to fix the issue you referenced could possibly increase the time to reach the 404 state. All it's doing is rewriting the URL displayed in the browser URL bar when the 404 state does get hit - all of the logic and lead up to getting to that 404 page, which is what's going to take measurable time, is untouched. The follow-up PR that just got merged some hours ago (after this issue was opened) makes some slight changes to the routing logic, but nothing that should measurably increase the loading time, and in any case those changes do not affect an application path like `/does/not/exist` - that logic is still untouched other than changing the displayed URL. Are you sure that the increased loading time observed here is not a general increase in prod-preview loading time? Perhaps some network change between you and the prod-preview server, or increased server load, is causing a false positive correlation between the load speed and the displayed URL change.

@jiekang yes, a path like `/does/not/exist` will fall through all of the routing cases and go to the `'**'` case, which will redirect to `_error` (404 page) without needing to make any additional network calls. A path like `/aazores@redhat.com/fake-space` (where `fake-space` doesn't exist) looks like a valid path to the router, so it will invoke the `ContextResolver` to check if the `aazores@redhat.com` user exists and if it has a `fake-space` space. So there will be these two chained requests made after the application has loaded, and the user will not be redirected to `_error` until after the final one fails. There's not much else we can do in this case however - the application cannot know that the path is actually invalid until it fails on requesting the space. Up until that point it looks like a completely valid path.

**Comment by christianvogt** [comment]
~54 requests and 1.7MB transferred in 5.3s before I got the 404 page with caching disabled.

There were 3 separate chunks of network requests that where each subsequent chunk of requests was blocked on the previous chunk first finishing. 

I'm not sure why we need to know the users information before determining if a route is valid or not but we currently are making requests (and duplicates) to our APIs.

If we could do routing before anything else, then we could probably cut down everything by 1/3 at least (some of those 54 requests were images and analytics).

As @andrewazores mentioned, if the first segments in the route do match an existing route it will be necessary to load that part of the app to determine if the subsequent route segments are valid or not. But again that part of the app should be able to determine if the route is valid quickly without loading and executing more code than necessary.

**Comment by mvulavak** [comment]
@christianvogt Is there a fix for this issue?

**Comment by christianvogt** [comment]
@mvulavak no there is no fix yet for this.

However the same test is now 39 requests, 1.4MB transferred.

The 404 shows up in a few seconds depending on the network speed and amount of cached files present on the user's machine. Never experienced the 30secs stated in the summary.

404s are edge cases and as such this issue isn't a high priority. We certainly could do better. It's worth keeping this open as we will be changing the main navigation soon which may affect this use case.
